src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransactionStore.cs

```diff
@@ -0,0 +1,216 @@
+    {
+        /// <summary>
+        /// Persist the template transaction into the database.
+        /// </summary>
+        /// <param name="templateTransaction">Template transaction to be inserted.</param>
+        Task PutAsync(TemplateTransaction templateTransaction);
+
+        /// <summary>
+        /// Get the template transaction from the database, identified by a hash.
+        /// </summary>
+        /// <param name="hash">A template transaction hash.</param>
+        /// <returns>The template transaction identified by the hash (or null if not found).</returns>
+        Task<TemplateTransaction> GetAsync(uint256 hash);
+
+        /// <summary>
+        /// Delete a template transaction from the database, identified by a hash.
+        /// </summary>
+        /// <param name="hash">Hash of the template transaction to be deleted.</param>
+        Task DeleteAsync(uint256 hash);
+
+        /// <summary>
+        /// Determine if a template transaction already exists in the database.
+        /// </summary>
+        /// <param name="hash">The hash of the template transaction.</param>
+        /// <returns><c>true</c> if the template transaction can be found in the database, otherwise return <c>false</c>.</returns>
+        Task<bool> ExistAsync(uint256 hash);
+    }
+
+    public class TemplateTransactionStore : ITemplateTransactionStore
+    {
+        private const string TemplateTransactionTableName = "TemplateTransaction";
+
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        /// <summary>Access to DBreeze database.</summary>
+        private readonly DBreezeEngine DBreeze;
+
+        private readonly Network network;
+
+        /// <summary>Provider of time functions.</summary>
+        private readonly IDateTimeProvider dateTimeProvider;
+
+        public TemplateTransactionStore(Network network, DataFolder dataFolder, IDateTimeProvider dateTimeProvider, ILoggerFactory loggerFactory)
+            : this(network, Path.Combine(dataFolder.RootPath, "sidechaindata"), dateTimeProvider, loggerFactory)
+        {
+        }
+
+        public TemplateTransactionStore(Network network, string folder, IDateTimeProvider dateTimeProvider, ILoggerFactory loggerFactory)
+        {
+            Guard.NotNull(network, nameof(network));
+            Guard.NotEmpty(folder, nameof(folder));
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+
+            Directory.CreateDirectory(folder);
+            this.DBreeze = new DBreezeEngine(folder);
+            this.network = network;
+            this.dateTimeProvider = dateTimeProvider;
+        }
+
+        /// <summary>Performs any needed initialisation for the database.</summary>
+        public virtual Task InitializeAsync()
+        {
+            Task task = Task.Run(() =>
+            {
+                this.logger.LogTrace("()");
+
+                // Currently don't do anything on startup
```

Comment:
Doesn't anything need to be initialised here ? current tip ? (I'm not fully sure what the template transaction does, but the common store pattern in the full node was to keep the current block hash height - maybe it's not need here).